crates/cranelift/src/func_environ.rs

```diff
@@ -2215,4 +2290,110 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
     fn use_x86_pmaddubsw_for_dot(&self) -> bool {
         self.isa.has_x86_pmaddubsw_lowering()
     }
+
+    cfg_if! {
+        if #[cfg(feature = "valgrind")] {
+            fn handle_before_return(
+                &mut self,
+                retvals: &[Value],
+                builder: &mut FunctionBuilder,
+            ) {
+                if self.valgrind {
+                    let func_index = match &builder.func.name {
+                        UserFuncName::User(user) => {
+                            FuncIndex::from_u32(user.index)
+                        }
+                        _ => {
+                            unreachable!()
```

Comment:
Let's change this to a `panic` and add a helpful message -- "function name not a `UserFuncName::User` as expected" or something like that?